core/src/test/java/org/apache/iceberg/TestLocationProvider.java

```diff
@@ -285,4 +286,22 @@ public void testObjectStorageWithinTableLocation() {
     assertThat(parts).element(2).asString().isNotEmpty();
     assertThat(parts).element(3).asString().isEqualTo("test.parquet");
   }
+
+  @TestTemplate
+  public void testEncodedFieldNameInPartitionPath() {
+    // Update the table to use a string field for partitioning with special characters in the name
+    table.updateProperties().set(TableProperties.OBJECT_STORE_ENABLED, "true").commit();
+    table.updateSchema().addColumn("data#1", Types.StringType.get()).commit();
+    table.updateSpec().addField("data#1").commit();
```

Comment:
Could we add a test for some other atypical characters like "$" or "&"? Some character from the "Characters that might require special handling" section in https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html.

Reply:
They recommend URL encoding the values, which is what we're doing here.  I'm not sure how much value there is in using a range of characters since we just validating that url encoding is happening.

The only one that immediately stands out to me is `=`, which does not get encoded, but that's been used for partition paths for a long time with both FileIO and Hadoop FileSystem.

Reply:
Maybe nitpicking, but in my reading of the AWS doc linked above I do not think they are recommending URL encoding. The doc's language is pretty vague. In fact, I suspect URL encoding even partition key values here will cause incorrect processing in tools that use `java.lang.URI` for parsing those locations in Iceberg metadata files because the escaped chars will be converted to their proper Unicode chars during parsing and will likely not match existing S3 keys when passed to `software.amazon.awssdk.services.s3.S3Utilities.parseUri(URI uri)`... I'm not saying there's definitely a bug there, but I strongly suspect interoperability issues.

Reply:
Example:
```
  @Test
  void testA() {
    URI uri = URI.create("s3://bucket/path/id%22=1/something.parquet");
    S3Utilities u = S3Utilities.builder().region(Region.EU_CENTRAL_1).build();
    S3Uri s3Uri = u.parseUri(uri);
    soft.assertThat(s3Uri.key().get()).isEqualTo("path/id%22=1/something.parquet");
  }
```
Outcome:
```
Expected :"path/id%22=1/something.parquet"
Actual   :"path/id"=1/something.parquet"
```
Note that Iceberg, by virtue of having its own `S3URI` class, will use `path/id%22=1/something.parquet` (with the `%` sign) as the S3 key at the S3 API level.

Reply:
For now, we'll just make this consistent with what we're doing for the values.  I agree we should probably up-level this discussion and figure out a path forward that is more usable with other URI parsing utilities.